ecl/hqlcpp/hqlres.cpp

```diff
@@ -288,10 +360,54 @@ void ResourceManager::addManifestsFromArchive(IPropertyTree *archive)
                 }
                 else
                 {
-                    VStringBuffer xpath("AdditionalFiles/Resource[@originalFilename=\"%s\"]", filename);
                     MemoryBuffer content;
+                    VStringBuffer xpath("AdditionalFiles/Resource[@originalFilename=\"%s\"]", filename);
+                    const char *md5=item.queryProp("@md5");
+                    if (!archive->hasProp(xpath.str()))
+                    {
+                        if (md5)
+                            xpath.clear().appendf("AdditionalFiles/Resource[@filename='%s'][@md5='%s']", filename, md5);
+                        else
+                            xpath.clear().appendf("AdditionalFiles/Resource[@originalFilename=\"%s\"]", filename);
+                        if (!archive->hasProp(xpath.str()))
+                            throw makeStringExceptionV(0, "Failed to locate resource for %s in archive", filename);
+                    }
                     archive->getPropBin(xpath.str(), content);
-                    addCompress(item.queryProp("@type"), content.length(), content.toByteArray(), &item);
+                    if (md5)
+                    {
+                        StringBuffer calculated;
+                        md5_data(content, calculated);
+                        if (!strieq(calculated, md5))
+                            throw makeStringExceptionV(0, "MD5 mismatch %s in archive", filename);
+                    }
+                    const char *type = item.queryProp("@type");
+                    if (strieq(type, "CPP") || strieq(type, "C"))
+                    {
+                        if (!ctxCallback->allowAccess("cpp", isSigned))
+                            throw makeStringExceptionV(0, "Embedded code via manifest file not allowed");
+                        if (!tempDir.length())
+                        {
+                            getTempFilePath(tempDir, "eclcc", nullptr);
+                            tempDir.append(PATHSEPCHAR).append("tmp.XXXXXX"); // Note - we share same temp dir for all from this manifest
+                            if (!mkdtemp((char *) tempDir.str()))
+                                throw makeStringExceptionV(0, "Failed to create temporary directory %s (error %d)", tempDir.str(), errno);
+                            cppInstance.addTemporaryDir(tempDir.str());
+                        }
+                        StringBuffer tempFileName;
+                        tempFileName.append(tempDir).append(PATHSEPCHAR).append(item.queryProp("@filename"));
+                        if (!recursiveCreateDirectoryForFile(tempFileName))
+                            throw makeStringExceptionV(0, "Failed to create temporary file %s (error %d)", tempFileName.str(), errno);
+                        FILE *source = fopen(tempFileName.str(), "wt");
```

Comment:
I'm not sure you need/want the "t" here.

Reply:
It's a c++ source file, which I would have thought was text. I very much doubt any compiler will care, but...